docs/reconfigurator.adoc

```diff
@@ -0,0 +1,183 @@
+
+The process of executing a blueprint involves a bunch of database queries and API calls to Sled Agent, Service Processors (via MGS), Dendrite, etc.  As a result:
+
+* Executing a blueprint is not atomic.
+* Executing a blueprint can take a while.
+* Any of the individual actions required to execute a blueprint can fail.
+* The component executing the blueprint (Nexus) can itself fail, either transiently or permanently.
+* To deal with that last problem, we have multiple Nexus instances.  But that means it's conceivable that multiple Nexus instances attempt to execute different blueprints concurrently or attempt to take conflicting actions to execute the same blueprint.
+
+=== The primacy of planning
+
+To make all this work:
+
+* Blueprints are immutable.  (If Nexus wants to change one, it creates a new one based on that one instead.)
+* Blueprints must be specific enough that there are **no meaningful choices for Nexus to make at execution time**.  For example, when provisioning a new zone, the blueprint specifies the new zone's id, what sled it's on, its IP address, and every other detail needed to provision it.
+* CockroachDB is the source of truth for which blueprint is the current target.
+* Every blueprint has a **parent blueprint**. A blueprint can only be made the target if its parent is the current target.
+
+The planner is synchronous, mostly deterministic, relatively simple, and highly testable.  This approach essentially moves all coordination among Nexus instances into the planning step.  Put differently: Nexus instances can generate blueprints independently, but only one will become the target, and that one is always an incremental step from the previous target.  Many tasks that would be hard to do in a distributed way (like allocating IPs or enforcing constraints like "no more than one CockroachDB node may be down for update at once") can be reduced to pretty straightforward, highly testable planner logic.
+
+As a consequence of all this:
+
+* At any given time, any Nexus instance may generate a new blueprint and make it the new system target (subject to the constraints above).  Multiple Nexus instances can generate blueprints concurrently.  They can also attempt to set the target concurrently.  CockroachDB's strong consistency ensures that only one blueprint can be the target at any time.
+* At any given time, any Nexus instance may be attempting to execute (realize) a blueprint that it believes is the latest target.  It may no longer be the current target, though.
```

Comment:
Maybe mention that the complexities are described in the section below. I wrote a comment then deleted it after realizing it's all discussed below. Yes, I should have known better ;)

Reply:
Done in 4bed612b900bb556e8cd3ec97d2394ea8d8f0ec5.